utils/build-script-utils/src/git.rs

```diff
@@ -0,0 +1,123 @@
+// Copyright 2019-2020 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+use std::{env, fs, fs::File, io, io::Read, path::PathBuf};
+
+/// Make sure the calling `build.rs` script is rerun when `.git/HEAD` changed.
+///
+/// The file is searched from the `CARGO_MANIFEST_DIR` upwards. If the file can not be found,
+/// a warning is generated.
+pub fn rerun_if_git_head_changed() {
+	let mut manifest_dir = PathBuf::from(
+		env::var("CARGO_MANIFEST_DIR").expect("`CARGO_MANIFEST_DIR` is always set by cargo."),
+	);
+	let manifest_dir_copy = manifest_dir.clone();
+
+	while manifest_dir.parent().is_some() {
+		match get_git_paths(&manifest_dir) {
+			Err(err) => {
+				eprintln!("cargo:warning=Unable to read the Git repository: {}", err);
+
+				return;
+			}
+			Ok(None) => {}
+			Ok(Some(paths)) => {
+				for p in paths {
+					println!("cargo:rerun-if-changed={}", p.display());
+				}
+
+				return;
+			}
+		}
+
+		manifest_dir.pop();
+	}
+
+	println!(
+		"cargo:warning=Could not find `.git/HEAD` searching from `{}` upwards!",
+		manifest_dir_copy.display(),
+	);
+}
+
+// Code taken from https://github.com/rustyhorde/vergen/blob/8d522db8c8e16e26c0fc9ea8e6b0247cbf5cca84/src/output/envvar.rs
```

Comment:
We do it on our own at the moment AND we don't use the code from the crate! See here: https://github.com/paritytech/substrate/blob/master/utils/build-script-utils/src/lib.rs#L25-L43

I mixed the 2 implementations because the one of vergen is actually better because it handles git worktree and our was also better because it handled cargo workspaces.

It is not possible to use the one in the crate as this code is private. It won't mix.